aip/general/0216.md

```diff
@@ -123,6 +123,44 @@ message PublishBookRequest {
 - The comment for the field **should** document the resource pattern.
 - Other fields **may** be included.
 
+### Declarative-friendly state transitions
+
+If a resource's state can be modified by [declarative tooling][aip-128], it
+**should** expose a separate `desired_state` field taking the same enum type.
+
+```proto
+message Book {
+  enum State {
+    STATE_UNSPECIFIED = 0;
+    UNPUBLISHED = 1;
+    PUBLISHING = 2;
+    PUBLISHED = 3;
+  }
+  
+  State state = 1 [(google.api.field_behaviopr = OUTPUT_ONLY)];
+  State desired_state = 2;
```

Comment:
Issues with this:

1. Many State values (e.g., transitional ones) wouldn't be valid for the desired state.
2. Enums can't be extended without potentially breaking clients.
3. Per-service state enums prevent cross-service lifecycle management. This is a major gap in AIP 152 (Jobs), as one example. Also in 164 (soft delete).

Separate attributes are often better. For instance, Kubernetes Deployment has a "paused" field:
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#deploymentspec-v1-apps

Reply:
1. As mentioned below, transitional states would be `INVALID_ARGUMENT` if supplied to `desired_state`.
2. I feel like I've known what scenario this breaks before but can't recall it - can you give an example?
3. It sounds like you're saying that this entire AIP is a mistake and should be deprecated -- if that's the case we really need guidance to replace it.

Reply:
> I feel like I've known what scenario this breaks before but can't recall it - can you give an example?

@bgrant0607 I feel like we should just update google.aip.dev/180 to explain this one. One reason is that clients unaware of the new enum value will receive the unspecified value, and send that over the wire potentially erasing the currently set parameter.

> It sounds like you're saying that this entire AIP is a mistake and should be deprecated -- if that's the case we really need guidance to replace it.

I'm not sure if it's that strong of a case, but we do need updated guidance that enables known gaps like cross-service lifecycle management. 

I definitely agree guidance needs be updated, but it's on a longer backlog. At minimum I think anything that removes guidance from AIP-128 and reduces the gap between declarative-friendly and non-declarative-friendly APIs is an improvement. Thanks for authoring this!

Reply:
To clarify I think Brian is arguing for more granular switches on fields (`paused` with a boolean vs `desired_state` that has the same set of possible enum values). I do agree that, at minimum, `desired_state` shouldn't have the ability to be set to transient states (`RUNNING`)

Reply:
I'd suggest this actually isn't a recommendation we should make. It may even be an anti-pattern, although it's understandable why it comes up as an idea every so often.

* It's true that custom methods are essentially invisible to desired-state systems

* It's also true that many some custom methods have the effect of having the output only state enum eventually reach a similarly-named value.

So it seems like there's a connection between the two. But having an assignable field of the output-only enum is very rarely the best way to design the new hand-free resource control feature.

A different question to ask is "What feature on this resource would I add that, when optionally configured by the customer, would tell the backend when these custom methods should be called automatically?"

Looking at VM instance, some custom methods that can't be reached are start/stop/reset/suspend/resume... A quick hypothetical of those...

```
message Instance {
  ...
  // Assign PowerControl details to enable automated control of running state.
  // Default behavior when nil is to leave all running-state to imperative method calls.
  PowerControl power_control;

  message PowerControl {
    // When specified, indicates that the appropriate custom method should be called as needed.
    // Calls to the custom methods "stop" "start" "suspend" and "resume" will fail if they contradict
    // the configured goal.
    enum goal = UNSPECIFIED | STARTED | SUSPENDED | STOPPED;

    // If the instance has been running since before this time, and it is not after this time,
    // then the instance will restart. Setting it to "now" is an effective way to cause a single reset.
    // Setting to a future time will schedule a single reset for when that time arrives.
    Timestamp restart_when;

    // other control features added here as they become necessary
  }
}
```

... just a gist, but you get the idea.

* if unassigned the current behavior (non-automated, custom method driven only) is fully in effect
* if there is an enum goal, the values can be only what's needed don't even necessarily have to align with actual output_only state positions
* any other resource-type-specific subtlety about rules and conditions of when those automatic effects can/should take place can be taken into account with related fields in the control structure